File: bindings/java/src/test/java/org/wickra/shazam/ShazamTest.java
package org.wickra.shazam;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class ShazamTest {
    private static final String SPEC =
            "{\"features\":[{\"kind\":\"price\",\"field\":\"close\"}],"
                    + "\"window\":1,\"metric\":\"euclid\"}";

    private static String candle(int time, int close) {
        return "{\"time\":" + time + ",\"open\":" + close + ",\"high\":" + close
                + ",\"low\":" + close + ",\"close\":" + close + ",\"volume\":1}";
    }

    @Test
    void versionIsNonEmpty() {
        assertFalse(Shazam.version().isEmpty());
    }

    @Test
    void indexAndMatchReturnsNearestTimestamp() {
        try (Shazam shazam = new Shazam(SPEC)) {
            StringBuilder history = new StringBuilder("[");
            for (int i = 1; i <= 10; i++) {
                if (i > 1) {
                    history.append(',');
                }
                history.append(candle(i, 100 + i));
            }
            history.append(']');

            String indexed = shazam.command("{\"cmd\":\"index\",\"history\":" + history + "}");
            assertTrue(indexed.contains("\"indexed\":10"), indexed);

            String raw = shazam.command(
                    "{\"cmd\":\"match\",\"current\":[" + candle(11, 110) + "],\"k\":3}");
            assertTrue(raw.contains("\"indexed\":10"), raw);
            // The tail of a monotone history is nearest its latest bar (ts 10).
            assertTrue(raw.contains("\"ts\":10"), raw);
        }
    }

    @Test
    void invalidSpecThrows() {
        assertThrows(IllegalArgumentException.class, () -> new Shazam("not json"));
    }

    @Test
    void unknownCommandIsInBandError() {
        try (Shazam shazam = new Shazam(SPEC)) {
            // An unknown command is not a hard error: the ABI returns a length and
            // the error surfaces in-band as {"ok":false,...} JSON.
            String raw = shazam.command("{\"cmd\":\"nope\"}");
            assertEquals(true, raw.contains("\"ok\":false"), raw);
        }
    }
}

File: bindings/node/__tests__/shazam.test.js
"use strict";

const { test } = require("node:test");
const assert = require("node:assert");
const { Shazam, version } = require("../index.js");

const SPEC = JSON.stringify({
  features: [{ kind: "price", field: "close" }],
  window: 1,
  metric: "euclid",
});

function candle(time, close) {
  return { time, open: close, high: close, low: close, close, volume: 1.0 };
}

test("index and match returns the nearest historical timestamp", () => {
  const shazam = new Shazam(SPEC);
  const history = Array.from({ length: 10 }, (_, i) =>
    candle(i + 1, 100 + i + 1),
  );
  const indexed = JSON.parse(
    shazam.command(JSON.stringify({ cmd: "index", history })),
  );
  assert.strictEqual(indexed.indexed, 10);

  const response = shazam.command(
    JSON.stringify({ cmd: "match", current: [candle(11, 110.0)], k: 3 }),
  );
  const report = JSON.parse(response);
  assert.strictEqual(report.indexed, 10);
  assert.strictEqual(report.matches.length, 3);
  assert.strictEqual(report.matches[0].ts, 10);
});

test("version matches the module-level function", () => {
  const shazam = new Shazam(SPEC);
  assert.strictEqual(shazam.version(), version());
});

test("a malformed spec throws", () => {
  assert.throws(() => new Shazam("not json"));
});
